### DISCUSSION TITLE
Add Rust-based Istanbul code instrumentation

### DISCUSSION THREAD

**Discussion by nicholaschiang** [vercel/next.js#30529] [Ideas]
### Describe the feature you'd like to request

I currently use [`babel-plugin-istanbul`](https://github.com/istanbuljs/babel-plugin-istanbul) to instrument my code to track code coverage while testing with [Cypress](https://docs.cypress.io/guides/tooling/code-coverage#Instrumenting-code):

```javascript
// babel.config.js
/**
 * Instruments our code (only when running tests).
 * @see {@link https://babeljs.io/docs/en/configuration#babelconfigjson}
 * @see {@link https://docs.cypress.io/guides/tooling/code-coverage.html#Instrumenting-code}
 */
module.exports = {
  presets: ['next/babel'],
  plugins: process.env.NODE_ENV !== 'production' ? ['istanbul'] : [],
};
```

### Describe the solution you'd like

I'd really like to use the new Rust-based compiler included by default in Next.js 12, but I can't until I'm able to instrument my code with Istanbul to track code coverage while testing.

### Describe alternatives you've considered

Another possible solution would Next.js providing an option to use the custom `babel.config.js` only while `NODE_ENV=test` or something similar (i.e. I can still take advantage of the faster build times with SWC while deploying to production on Vercel but still use Istanbul to track code coverage while testing on GitHub Actions).

**Comment by timneutkens** [comment]
Please use https://github.com/vercel/next.js/discussions/30174 for sharing what Babel transforms are being used, that way it's all kept in a single place.

**Reply by eden-chan** [reply]
Hacky solutions found here 
https://stackoverflow.com/questions/71564063/can-i-get-next-js-to-exclude-babelrc-during-build

**Comment by LarsFlieger** [comment]
One more reason to have this feature is when your code uses exclusive SWC features like `@next/font` ([see problem](https://nextjs.org/docs/messages/babel-font-loader-conflict)). You are not able to instrument your code anymore.

**Reply by vjpr** [reply]
Try https://github.com/kwonoj/swc-plugin-coverage-instrument

**Reply by sgoodluck** [reply]
It isn't clear to me at all how that plugin is supposed to work. The documentation needs some expanding. Do you have a working example?

**Reply by IonelLupu** [reply]
anyone figured how https://github.com/kwonoj/swc-plugin-coverage-instrument should work?

**Comment by alamenai** [comment]
May this feature is related to this [babel plugin with swc](https://github.com/istanbuljs/babel-plugin-istanbul/issues/285)

**Comment by mkschreder** [comment]
Having the same problem. It would seem coverage would be one of the most basic features absolutely essential to have in place, but no...

**Comment by h2oearth** [comment]
Do you have any news about this feature? I tried this [plugin](https://github.com/kwonoj/swc-plugin-coverage-instrument), but it is broken  :-/ . We use SWC and babel is not an option !

**Reply by IonelLupu** [reply]
hey. Did you figure out show that plugin works?

**Reply by h2oearth** [reply]
I give up because we are using the CY cloud, and they are about to add code coverage to their cloud platform.

**Comment by bashlk** [comment]
After some messing about, I was able to get `swc-plugin-coverage-instrument` to work in a Next.js 12 project that uses SWC. Cypress was then able to use the instrumented build it generated to generate a coverage report.

Here is what worked for me
- Install `swc-plugin-coverage-instrument@0.0.12` by running `yarn add -D swc-plugin-coverage-instrument@0.0.12` or its equivalent for your package manager. Using version 0.0.12 since newer versions error out alongside Next 12.3.4. (See: https://github.com/kwonoj/swc-plugin-coverage-instrument/issues/180)  If you have a newer Next.js version, the newer versions might work.
- Modify Next.js's SWC config to use the plugin
```
const nextConfig = {
   ....
   experimental: {
    swcPlugins: [
      [
        "swc-plugin-coverage-instrument",
        {
          coverageVariable: "__coverage__",
        },
      ],
    ],
  },
}
```
- Follow this section of the cypress docs to set it up to use the instrumented builds - https://docs.cypress.io/guides/tooling/code-coverage#E2E-code-coverage
- `npx next build` and then `npx next start` alongside `npx cypress run` (or whatever scripts you have in package.json for invoking these commands)

Once cypress has finished running all the specs, a coverage report will be placed in a folder named `coverage` in the root of your project.

**Comment by RobertFent** [comment]
Any news? We are using Nextjs v.14.2.28 and I cant get it to work. Using Babel is no option
